src/apps_plugin.c

```diff
@@ -1760,20 +1760,20 @@ static inline int read_pid_file_descriptors(struct pid_stat *p, void *ptr) {
             p->fds_size = fdid + MAX_SPARE_FDS;
         }
 
-        if(unlikely(p->fds[fdid] == 0)) {
-            // we don't know this fd, get it
+        // get fd information
+        sprintf(fdname, "%s/proc/%d/fd/%s", netdata_configured_host_prefix, p->pid, de->d_name);
```

Comment:
hm... this means on a system with many thousands of fds, we will do many thousands of sprintfs per second. Have you checked the increased CPU utilization due to this?

I see `readdir()` returns the inode number. We could probably cache that inode number and compare it at the next iteration.

If this works, it will save the `sprintf()` and the `readlink()`.

What do you think?

Reply:
I agree. In the current implementation there must be indeed some impact of calling `readlink()` on every single fd for each collection.

I'll try adding an array of inodes to each pid_stat then, much like we currently do with the `fdid->slot` assignation.

Reply:
Can you please check using your test tool this will work?
It may not report inodes or something.

Also, it may be easier to convert the existing array of ints to an array of structures, with each structure having 2 ints (fd, inode). I can't tell right now which way is better.

Reply:
Actually, on second though, I'm having some doubts about this. It has been already specified here: https://github.com/firehol/netdata/blob/5886a64/src/apps_plugin.c#L1787

When a FD number is reused for another link, the inode number stays the same. So we'll always pass the inode check against the cache, even if the link type isn't the same.

Reply:
hm... this means the best we can do is to cache the filenames, so that we will not do `sprintf` all the time. This will be tricky, since we will have to `strdupz()` and `freez()` these buffers.

Then, we may be able to save a bit on resources, if we don't `readlink()` every single second. So, we could have a timestamp too and execute `readlink()` every 2, 3, or 5 seconds, depending on a global option.